23 informers deliverable

emmjohnson opened the pull request.

Commit:
Add integration test for deliverable informers

Commit:
Add dynamic informers and template watchers

Commit:
Add unit tests

Commit:
Lint

Commit:
better aliases for deliverable fakes

squeedee reviewed (changes requested):
We shouldn't review this until #304 lands

Also we've got a flake that should have died by now, will investigate:
![Screen Shot 2021-11-08 at 11 50 09 PM](https://user-images.githubusercontent.com/27838/140864410-83de7d26-4933-4325-9835-01324a2f2342.png)

Commit:
Remove backoff tests

- Backofftests made sense when we used errors and timeouts to ensure
  reconcile was called again.
- Now when a reconcile error occurs, backoff will be shortcircuited
  by informers. Testing informers work correctly is all we need.

[#23]

Co-authored-by: Emily Johnson <emjohnson@vmware.com>

Commit:
Fix flaky test

- We're dependent on APIServer + our countroller resolving this stamp
  in under a second, which just doesn't always happen.

Co-authored-by: Emily Johnson <emjohnson@vmware.com>

Commit:
Fix hard to read inner range
[#23]

Co-authored-by: Emily Johnson <emjohnson@vmware.com>

martyspiewak commented on pkg/controller/deliverable/reconciler.go:
```diff
@@ -31,25 +34,38 @@ import (
+type Reconciler interface {
+	Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error)
+	AddTracking(dynamicTracker DynamicTracker)
+}
```
This interface is declared here, in workload, and in runnable. I think it would make more sense to declare it once where we use it, in the registrar, rather than redeclaring it in each package.

cirocosta commented on pkg/registrar/registrar.go:
```diff
@@ -213,13 +213,18 @@ func registerDeliverableController(mgr manager.Manager) error {
 		return fmt.Errorf("controller new: %w", err)
 	}
 
+	reconciler.AddTracking(&external.ObjectTracker{
```
one implicit detail of using `external.ObjectTracker` is that one can actually "pause the watching" by making use of clusterapi's `paused` annotation:

```go
          // PausedAnnotation is an annotation that can be applied to any Cluster API
          // object to prevent a controller from processing a resource.
          //
          // Controllers working with Cluster API objects must check the existence of this annotation
          // on the reconciled object.
          PausedAnnotation = "cluster.x-k8s.io/paused"
```

used as a predicate in

```go
          log.Info("Adding watcher on external object", "GroupVersionKind", gvk.String())
          err := o.Controller.Watch(
                  &source.Kind{Type: u},
                  handler,
                  predicates.ResourceNotPaused(log),
          )
```

(_see https://github.com/kubernetes-sigs/cluster-api/blob/6abfc64139ade5a225ba62d675117c510885f1e2/controllers/external/tracker.go#L56-L60_)

emmjohnson replied:
Interesting...

Commit:
Remove Reconciler interface
